extensions/amp-ad/0.1/amp-ad-api-handler.js

```diff
@@ -121,21 +122,36 @@ export class AmpAdApiHandler {
       // creative as it will not expect having to send the render-start message.
       this.iframe_.style.visibility = 'hidden';
     }
-    listenForOnce(this.iframe_, 'render-start', () => {
-      if (!this.iframe_) {
-        return;
+
+    const promise = new Promise(resolve => {
+      listenForOnce(this.iframe_, 'render-start', () => {
+        resolve();
+      }, this.is3p_);
+    });
+
+    this.renderStartPromise_ = timerFor(this.baseInstance_.win).timeoutPromise(
+        2000, promise,
+        'fail to receive render-start event from ad server, timeout');
+
+    this.renderStartPromise_.then(() => {
+      //TODO: report performance
+      if (this.iframe_) {
+        this.iframe_.style.visibility = '';
       }
-      if (this.baseInstance_.renderStartResolve_) {
-        this.baseInstance_.renderStartResolve_();
-        this.baseInstance_.renderStartResolve_ = null;
+    }).catch(() => {
+      //TODO: report performance
+      if (this.iframe_) {
+        this.iframe_.style.visibility = '';
       }
-      this.iframe_.style.visibility = '';
-    }, this.is3p_);
+    });
+
     this.viewer_.onVisibilityChanged(() => {
       this.sendEmbedInfo_(this.baseInstance_.isInViewport());
     });
     this.element_.appendChild(this.iframe_);
-    return loadPromise(this.iframe_);
+    return loadPromise(this.iframe_).then(() => {
```

Comment:
Since we want to create `timeoutPromise` after this.iframe_ is loaded, I can't do return promise.all here.